css/css-nesting/parsing.html

```diff
@@ -39,30 +39,48 @@
     }, "INVALID: " + ruleText);
   }
 
+  // basic usage
   testNestedSelector("&");
   testNestedSelector("&.bar");
   testNestedSelector("& .bar");
   testNestedSelector("& > .bar");
-  testNestedSelector("> & .bar");
+
+  // relative selector
   testNestedSelector("> .bar", {expected:"& > .bar"});
+  testNestedSelector("> & .bar", {expected: "& > & .bar"});
   testNestedSelector("+ .bar &", {expected:"& + .bar &"});
-  testNestedSelector(".test > & .bar");
   testNestedSelector("+ .bar, .foo, > .baz", {expected:"& + .bar, & .foo, & > .baz"});
-  testNestedSelector(".foo, .foo &", {expected:"& .foo, .foo &"});
+
+  // implicit relative (and not)
+  testNestedSelector(".foo");
```

Comment:
@tabatkins Why do we not expect `& .foo` here? I see nothing in the recent edit that does anything special for relative selectors with a descendant combinator? https://drafts.csswg.org/css-nesting-1/#cssom

Reply:
@andruud I'm confused, my understanding is that these are not really relative selectors.

I guess they do match the `<relative-selector>` grammar without a combinator, tho... But you can't really serialize `&` unconditionally for all those, because then `.foo &` would become `& .foo &`, which is not really equivalent / what you want, right?

Reply:
@emilio: `.foo` is a relative selector, but `.foo &` is not a relative selector, see the first two bullet points here: https://drafts.csswg.org/css-nesting-1/#syntax

Reply:
Ah, I see. Well both are relative selectors per syntax, and none have an extra combinator, but yeah, you're right.

I don't think we want to serialize the implied combinator, since not having it doesn't make the selector invalid in any other context (unlike for relative selectors that actually start with a combinator), but I wouldn't mind either way I guess, if you have a strong reason to prefer serializing the `&` there...

Reply:
The most consistent is to keep serializing as the shortest possible, right? So no &, like the spec was up until now.

Reply:
Well, but:

 * Impls were not serializing it at the shortest possible form (`& > foo` wasn't shortened to `> foo` for example).
 * (more importantly) the shortest form is not valid outside nesting contexts.

The later point is why I argued for serializing implicit rules and the `&` at least in relative selectors with a combinator.

Otherwise you don't have the guarantee that `ruleA.selectorText = ruleB.selectorText;` does something useful, for example (which IMO it always should).

Reply:
I can agree & > foo should have been serialized to > foo.

Demanding that selectorText is movable between different contexts is a new goal to me, and not one that I immediately think is important? I mean, & doesn't always do anything useful outside a nested context in the first place.

Reply:
I think it's important / a worth-preserving property that things generally round-trip.

That applies both for `.cssText` and `.selectorText`. It seems error-prone if something out of `.cssText` isn't guaranteed to be valid CSS depending on the context.

Anyways, if you strongly disagree with the resolution for https://github.com/w3c/csswg-drafts/issues/8970, it should be re-opened and re-discussed there (or in a new issue), not here...

Reply:
Yeah, "moving between contexts" isn't a *strong* goal, but it is a nice thing to have so the CSS you author is less context-specific. This was a weak argument in favor of the current nesting syntax overall, for example - it meant you wrote the same rules whether you were nested in a style rule or an `@scope` rule.

It's definitely something we can break if necessary, but if preserving this quality doesn't hurt other goals we should preserve it.

Reply:
OK, but my [original question](https://github.com/web-platform-tests/wpt/pull/41213#discussion_r1282796593) (that's still not answered) was about a detail of the existing resolution:

```
  // implicit relative (and not)
  testNestedSelector(".foo");
```

Shouldn't this expect `& .foo`?

Reply:
OK, given no response from @tabatkins: since the spec describes the behavior, the call notes doesn't say otherwise, and @emilio says he can live with it,  then I think we'll just change the WPT to expect that relative `.foo` is serialized as `& .foo`.

Reply:
I think it's slightly unfortunate, fwiw, but I can live with it yeah

Reply:
Oh sorry for missing this comment! 

I don't have much of an opinion on which way we go. You're right that, per a strict reading, the inner selector of `.foo { .bar {...}}` is a relative selector - nested style rules use `<relative-selector-list>` for their grammar, implying that their selectors are all "relative selectors", and only exempt one case from that interpretation (when the selector contains an explicit `&`). Then the CSSOM section does say that you must absolutize all relative selectors, inserting the implied `&`.

We added that serialization requirement to make sure that the selectors were valid when transferred to other contexts, and that reasoning doesn't have anything to say about a selector like ".foo" - both ".foo" and "& .foo" would be valid when moved around (and approximately equal in behavior, except that the latter can't select the root element).

Whether we fix this by adjusting the WPT, or by fixing the spec to only insert the `&` when it's needed to prevent the selector from starting with a combinator (and thus being invalid in non-nested contexts), I don't care too much.

Reply:
Minor preference for changing the spec, but can live with either really.

Reply:
I've changed the WPT to follow the spec here https://github.com/web-platform-tests/wpt/pull/42560 if we want to go in this direction